### PR TITLE
New module with integration tests common logic

### PR DESCRIPTION
**Pull request by hristobojilov**
[#173] New module with integration tests common logic

Signed-off-by: Hristo Bozhilov <Hristo.Bozhilov@bosch.io>

### CODE REVIEW

**Review by e-grigorov — CHANGES_REQUESTED** [review]
Could you please run all tools from the validation flow.

File: integration/testutil/web.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("%s %s request failed: %s", method, url, resp.Status)
+	}
+
+	return io.ReadAll(resp.Body)
+}
+
+// NewWSConnection creates new web socket connection
+func NewWSConnection(cfg *TestConfig) (*websocket.Conn, error) {
```
Comment (gboyvalenkov-bosch) [inline integration/testutil/web.go:49]: Is it planned to add listening for ditto events?

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: integration/util/config.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	LocalBroker              string `env:"LOCAL_BROKER" envDefault:"tcp://localhost:1883"`
+	MqttQuiesceMs            int    `env:"MQTT_QUIESCE_MS" envDefault:"500"`
+	MqttAcknowledgeTimeoutMs int    `env:"MQTT_QUIESCE_MS" envDefault:"3000"`
+
+	DigitalTwinAPIAddress  string `env:"DIGITAL_TWIN_API_ADDRESS"`
+	DigitalTwinAPIUserName string `env:"DIGITAL_TWIN_API_USERNAME" envDefault:"ditto"`
+	DigitalTwinAPIPassword string `env:"DIGITAL_TWIN_API_PASSWORD" envDefault:"ditto"`
+
+	EventTimeoutMs  int `env:"EVENT_TIMEOUT_MS" envDefault:"30000"`
+	StatusTimeoutMs int `env:"STATUS_TIMEOUT_MS" envDefault:"10000"`
```
Comment (e-grigorov) [inline integration/util/config.go:34]: Is it a common logic?

File: integration/util/config.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+	TimeDeltaMs int `env:"TIME_DELTA_MS" envDefault:"5000"`
```
Comment (e-grigorov) [inline integration/util/config.go:36]: Is it common logic?

File: integration/util/client.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+
+package testutil
```
Comment (e-grigorov) [inline integration/util/client.go:13]: Is it by mistake? Are we going to change the package name? Usually, folder name and the package name are aligned.

File: integration/util/client.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+// SendMessage sends a message to a topic using specified client
+func SendMessage(cfg *TestConfig, client MQTT.Client, topic string, message interface{}) error {
+	payload, err := json.Marshal(message)
+	if err != nil {
+		return err
+	}
+	token := client.Publish(topic, 1, false, payload)
+	timeout := time.Duration(cfg.MqttAcknowledgeTimeoutMs * int(time.Millisecond))
+	if !token.WaitTimeout(timeout) {
+		return ditto.ErrAcknowledgeTimeout
```
Comment (e-grigorov) [inline integration/util/client.go:53]: Is it ok to reference Ditto lib here? This method is pure MQTT.

File: integration/util/client.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+	if token := mqttClient.Connect(); token.Wait() && token.Error() != nil {
+		return nil, token.Error()
+	}
+
+	return mqttClient, nil
+}
+
+// SendMessage sends a message to a topic using specified client
+func SendMessage(cfg *TestConfig, client MQTT.Client, topic string, message interface{}) error {
```
Comment (e-grigorov) [inline integration/util/client.go:45]: We have NewMQTTClient, do we need to rename it to: SendMQTTMessage?

File: integration/util/suite.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+// SuiteInit is testify Suite initialization helper
+type SuiteInit struct {
+	Cfg *TestConfig
+
+	DittoClient *ditto.Client
+	MqttClient  mqtt.Client
+}
+
+// Setup established connections to local broker and Ditto
```
Comment (e-grigorov) [inline integration/util/suite.go:35]: - established -> establishes
- local broker -> local MQTT broker

File: integration/util/suite.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+func (suite *SuiteInit) Setup(t *testing.T) {
+	cfg := &TestConfig{}
+
+	opts := env.Options{RequiredIfNoDef: true}
+
+	if err := env.Parse(cfg, opts); err != nil {
+		t.Skip(err)
```
Comment (e-grigorov) [inline integration/util/suite.go:42]: I guess that TC has to fail, if environment variable can not be processed.

File: integration/util/web.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"golang.org/x/net/websocket"
+)
+
+// SendRequest sends new HTTP request to ditto REST API
+func SendRequest(cfg *TestConfig, method string, url string) ([]byte, error) {
```
Comment (e-grigorov) [inline integration/util/web.go:28]: SendRequest sounds generic.

File: integration/util/web.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+	if url.Scheme == "https" {
+		return fmt.Sprintf("wss://%s:%s", url.Hostname(), getPortOrDefault(url, "443")), nil
+	}
+
+	return fmt.Sprintf("ws://%s:%s", url.Hostname(), getPortOrDefault(url, "80")), nil
+}
+
+// BeginWait waits for an error to be received over the channel up to a timeout
+func BeginWait(timeout time.Duration, resultCh chan error, closer func()) chan error {
```
Comment (e-grigorov) [inline integration/util/web.go:93]: BeginWait - can be call it something like: WaitError? We pass channel and return channel. Are we sure about the usage?

File: integration/util/web.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+			closer()
+			ch <- errors.New("timeout")
+		}
+	}()
+
+	return ch
+}
+
+// BeginWSWait waits for a message to be received via websocket
+func BeginWSWait(cfg *TestConfig, ws *websocket.Conn, check func(payload []byte) error) chan error {
```
Comment (e-grigorov) [inline integration/util/web.go:110]: The name has to be aligned with the previous function.

File: integration/util/suite.go (hunk 8 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/suite.go:42]: Yes, please use t.FailNow. See the suite-connector integration test PR.

File: integration/util/web.go (hunk 10 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go:93]: The usage is: wait on the channel for the timeout duration. If there's no result after the given timeout, call the close function and send an error the the channel.

File: integration/util/web.go (hunk 9 above)
Comment (e-grigorov) [reply integration/util/web.go]: `SendDittoRequest` is going in the right direction. One note more, the environment variables are called `DigitalTwin...` i.e. the focus is more on the functional part. I would suggest to call this method: `SendDigitalTwinRequest`.

File: integration/util/web.go (hunk 10 above)
Comment (e-grigorov) [reply integration/util/web.go]: @dimiter-todorov-georgiev - my confusion is that it's an error channel i.e. it waits for an error or timeout. I still cannot completely understand the flow.

File: integration/util/web.go (hunk 11 above)
Comment (e-grigorov) [reply integration/util/web.go:110]: One candidate: `WaitWSMessage`

File: integration/util/web.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+	resultCh := make(chan error)
+
+	go func() {
+		var payload []byte
+		threshold := time.Now().Add(timeout)
+		var err error
+		for time.Now().Before(threshold) {
+			err = websocket.Message.Receive(ws, &payload)
+			if err == nil {
+				err = check(payload)
```
Comment (e-grigorov) [inline integration/util/web.go:121]: API has to be revised. Currently, payload is processed by `check` function and error channel was to be requested. It seems a little bit more complicated than it's should be just to read one message.

File: integration/util/client.go [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+	mqttClient := MQTT.NewClient(opts)
+
+	if token := mqttClient.Connect(); token.Wait() && token.Error() != nil {
+		return nil, token.Error()
+	}
+
+	return mqttClient, nil
+}
+
+// SendMQTTMessage sends a message to a topic using specified client
```
[the remark this answers is not in the document]
Comment (e-grigorov) [reply integration/util/client.go]: JSON Format -> JSON format

File: integration/util/client.go (hunk 5 above)
Comment (hristobojilov) [reply integration/util/client.go]: Looks ok to me

File: integration/util/web.go [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+				return nil
+			}
+		}
+	}
+
+	return errors.New("timeout")
+}
+
+// ProcessMessages polls messages from the web socket connection until specific condition is satisfied or timeout expires
+func ProcessMessages(
```
Comment (e-grigorov) [inline integration/util/web.go:121]: ProcessWSMessages?

File: integration/util/web.go [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+	if url.Scheme == "https" {
+		return fmt.Sprintf("wss://%s:%s", url.Hostname(), getPortOrDefault(url, "443")), nil
+	}
+
+	return fmt.Sprintf("ws://%s:%s", url.Hostname(), getPortOrDefault(url, "80")), nil
+}
+
+// WaitForAck polls messages from the web socket connection until specific acknowledgement is received or timeout expires
+func WaitForAck(
+	timeout time.Duration,
```
Comment (e-grigorov) [inline integration/util/web.go:98]: Can we have a common configuration about this? It'll be some kind of overhead for the integration tests.

File: integration/util/web.go [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+		var envelope *protocol.Envelope
+		if err == nil {
+			envelope = &protocol.Envelope{}
+			err = json.Unmarshal(payload, envelope)
+			if err == nil {
+				stop, err = process(envelope)
+			} else {
+				// Unmarshalling error, the payload is not a JSON of protocol.Envelope
+				// Ignore the error
+				fmt.Fprintf(os.Stderr, "error unmarshalling a protocol.Envelope: %v", err)
```
Comment (e-grigorov) [inline integration/util/web.go:143]: It's a test utility and cannot log on the standard error stream.

File: integration/util/web.go [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	return errors.New("timeout")
+}
+
+// ProcessMessages polls messages from the web socket connection until specific condition is satisfied or timeout expires
+func ProcessMessages(
+	timeout time.Duration,
+	ws *websocket.Conn,
+	process func(*protocol.Envelope) (bool, error)) (bool, error) {
```
Comment (e-grigorov) [inline integration/util/web.go:124]: What's the reason to return this bool param? FMPOV, there is an error or not.

File: integration/util/web.go (hunk 14 above)
Comment (e-grigorov) [inline integration/util/web.go:121]: Do we need this as a public method?

File: integration/util/web.go [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+
+	if url.Scheme == "https" {
+		return fmt.Sprintf("wss://%s:%s", url.Hostname(), getPortOrDefault(url, "443")), nil
+	}
+
+	return fmt.Sprintf("ws://%s:%s", url.Hostname(), getPortOrDefault(url, "80")), nil
+}
+
+// WaitForAck polls messages from the web socket connection until specific acknowledgement is received or timeout expires
+func WaitForAck(
```
Comment (e-grigorov) [inline integration/util/web.go:97]: Do we need this as a public method?

File: integration/util/web.go [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+			stopped: stopped,
+			err:     err,
+		}
+	}()
+
+	return responseCh
+}
+
+// WaitSubscribeResult waits until a SubscribeResult appears or timeout
+func WaitSubscribeResult(timeout time.Duration, resultCh chan SubscribeResult, closer func()) SubscribeResult {
```
Comment (e-grigorov) [inline integration/util/web.go:184]: `closer` is not used?

File: integration/util/web.go [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+// SubscribeResult contains subscription information
+type SubscribeResult struct {
+	stopped bool
+	err     error
+}
+
+// Subscribe starts ProcessMessages asynchronously
+func Subscribe(
```
Comment (e-grigorov) [inline integration/util/web.go:166]: We are in util package, revise all method names. These here should be WS specific.

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: integration/testutil/web.go (hunk 1 above)
Comment (e-grigorov) [reply integration/testutil/web.go]: Related API update is in progress.

File: integration/util/web.go (hunk 14 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go:121]: Yes, in case we want to process the messages synchronously.

File: integration/util/web.go (hunk 18 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go:97]: Yes, we need it in some test cases.

File: integration/util/web.go [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+
+	if stop {
+		return true, nil
+	}
+
+	return false, fmt.Errorf("WS response not received in %v, last error: %v", timeout, err)
+}
+
+// SubscribeResult contains subscription information
+type SubscribeResult struct {
```
Comment (e-grigorov) [inline integration/util/web.go:160]: Is it really a subscribe result? As I can understand it, it's more about the cancellation of the subscription.

File: integration/util/web.go (hunk 17 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go:124]: The idea is to separate "finished/stopped" from "an error has occurred".

File: integration/util/web.go (hunk 20 above)
Comment (e-grigorov) [inline integration/util/web.go:166]: How to unsubscribe? Just to close the WS connection? Do I need a special error there?

File: integration/util/web.go (hunk 10 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/web.go]: We have removed the async behavior as it wasn't required for our tests.

File: integration/util/mqtt_client.go [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+// SendMQTTMessage sends a message to a topic using specified client. The message is serialized to JSON Format.
+func SendMQTTMessage(cfg *TestConfiguration, client MQTT.Client, topic string, message interface{}) error {
+	payload, err := json.Marshal(message)
+	if err != nil {
+		return err
+	}
+	token := client.Publish(topic, 1, false, payload)
+	timeout := time.Duration(cfg.MqttAcknowledgeTimeoutMs * int(time.Millisecond))
```
Comment (e-grigorov) [inline integration/util/mqtt_client.go:51]: Is there a reason not to use the new utility method: MillisToDuration?

File: integration/util/mqtt_client.go [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+	if !token.WaitTimeout(timeout) {
+		return errors.New("timeout")
```
Comment (e-grigorov) [inline integration/util/mqtt_client.go:53]: The message can be more concrete. For example, `timeout while sending MQTT message`. Just discussion point, do we need to add message specific information?

File: integration/util/mqtt_client.go [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+			ch <- result{nil, err}
+		}
+		ch <- result{&cfg, nil}
+	}); token.Wait() && token.Error() != nil {
+		return nil, token.Error()
+	}
+
+	defer mqttClient.Unsubscribe("edge/thing/response")
+
+	if token := mqttClient.Publish("edge/thing/request", 1, false, ""); token.Wait() && token.Error() != nil {
```
Comment (e-grigorov) [inline integration/util/mqtt_client.go:86]: This method is full with wait without timeout. Is it ok or some kind of acknowledge timeout has to be used?

File: integration/util/suite.go [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+
+// SuiteInitializer  is testify Suite initialization helper
+type SuiteInitializer struct {
+	Cfg *TestConfiguration
+
+	DittoClient *ditto.Client
+	MqttClient  MQTT.Client
+}
+
+// Setup establishes connections to local MQTT broker and Ditto
```
Comment (e-grigorov) [inline integration/util/suite.go:35]: `to local MQTT broker` -> `to the local MQTT broker`

File: integration/util/mqtt_client.go [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+// GetThingConfiguration retrieves ThingConfig using specified client
+func GetThingConfiguration(mqttClient MQTT.Client) (*ThingConfiguration, error) {
+	type result struct {
+		cfg *ThingConfiguration
+		err error
+	}
+
+	ch := make(chan result)
+
+	if token := mqttClient.Subscribe("edge/thing/response", 1, func(client MQTT.Client, message MQTT.Message) {
```
Comment (gboyvalenkov-bosch) [inline integration/util/mqtt_client.go:74]: edge/thing/response should be a constant

File: integration/util/mqtt_client.go (hunk 24 above)
Comment (gboyvalenkov-bosch) [inline integration/util/mqtt_client.go:86]: use a constant for edge/thing/request

File: integration/util/suite.go [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+func (suite *SuiteInitializer) Setup(t *testing.T) {
+	cfg := &TestConfiguration{}
+
+	opts := env.Options{RequiredIfNoDef: true}
+	require.NoError(t, env.Parse(cfg, opts), "Failed to process environment variables")
```
Comment (e-grigorov) [inline integration/util/suite.go:40]: `Failed to` -> `failed to` - as it's for all other error messages.

File: integration/util/suite.go [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+	require.NoError(t, err, "connect to MQTT broker")
+
+	dittoClient, err := ditto.NewClientMQTT(mqttClient, ditto.NewConfiguration())
+	if err == nil {
+		err = dittoClient.Connect()
+	}
+
+	if err != nil {
+		mqttClient.Disconnect(uint(cfg.MqttQuiesceMs))
+		require.NoError(t, err, "initialize ditto client")
```
Comment (gboyvalenkov-bosch) [inline integration/util/suite.go:54]: just use require.Failf(...)

File: integration/util/suite.go [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+	mqttClient, err := NewMQTTClient(cfg)
+	require.NoError(t, err, "connect to MQTT broker")
+
+	dittoClient, err := ditto.NewClientMQTT(mqttClient, ditto.NewConfiguration())
+	if err == nil {
+		err = dittoClient.Connect()
+	}
+
+	if err != nil {
+		mqttClient.Disconnect(uint(cfg.MqttQuiesceMs))
```
Comment (gboyvalenkov-bosch) [inline integration/util/suite.go:53]: do not disconnect if connect had not been called

File: integration/util/web.go [hunk 30]
```diff
@@ -0,0 +1,172 @@
...
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return nil, fmt.Errorf("%s %s request failed: %s", method, url, resp.Status)
+	}
+
+	return io.ReadAll(resp.Body)
+}
+
+// NewWSConnection creates new web socket connection
+func NewWSConnection(cfg *TestConfiguration) (*websocket.Conn, error) {
```
Comment (e-grigorov) [inline integration/util/web.go:72]: Just a discussion point, NewDigitalTwinWSConnection - it's more correct but it seems too long as a name. What do you think?

File: integration/util/web.go [hunk 31]
```diff
@@ -0,0 +1,172 @@
...
+		if err != nil {
+			return fmt.Errorf("error reading from websocket: %s", err)
+		}
+		message := strings.TrimSpace(string(payload))
+		if message == expectedMessage {
+			return nil
+		}
+	}
+
+	return errors.New("timeout")
```
Comment (e-grigorov) [inline integration/util/web.go:131]: The message can be more concrete.

File: integration/util/web.go [hunk 32]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+// ProcessWSMessageResult is the result of messages processing
+type ProcessWSMessageResult struct {
```
Comment (e-grigorov) [inline integration/util/web.go:135]: No channels, no go routines any more, it seems that this struct is not needed any more or I am missing something?

File: integration/util/web.go [hunk 33]
```diff
@@ -0,0 +1,172 @@
...
+		unmarshalErr := json.Unmarshal(payload, envelope)
+		if unmarshalErr == nil {
+			result = process(envelope)
+		} else {
+			// Unmarshalling error, the payload is not a JSON of protocol.Envelope
+			// Ignore the error
+		}
+	}
+
+	if !result.Finished {
```
Comment (gboyvalenkov-bosch) [inline integration/util/web.go:167]: Maybe returning only an error is sufficient in this case

File: integration/util/web.go [hunk 34]
```diff
@@ -0,0 +1,172 @@
...
+
+	if url.Scheme == "https" {
+		return fmt.Sprintf("wss://%s:%s", url.Hostname(), getPortOrDefault(url, "443")), nil
+	}
+
+	return fmt.Sprintf("ws://%s:%s", url.Hostname(), getPortOrDefault(url, "80")), nil
+}
+
+// WaitForWSMessage polls messages from the web socket connection until specific message is received or timeout expires
+func WaitForWSMessage(timeout time.Duration, ws *websocket.Conn, expectedMessage string) error {
```
Comment (e-grigorov) [inline integration/util/web.go:115]: Do we need to unify the first argument. Most of the methods are using TestConfiguration, here, timeout is used.

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: integration/util/web.go (hunk 32 above)
Comment (gboyvalenkov-bosch) [reply integration/util/web.go:135]: It is useful for callback to _process_ func.

File: integration/util/mqtt_client.go (hunk 24 above)
Comment (dimiter-todorov-georgiev) [reply integration/util/mqtt_client.go:86]: We should use `WaitTimeout(time.Duration) bool` in both places in this function.

File: integration/util/web.go (hunk 33 above)
Comment (gboyvalenkov-bosch) [reply integration/util/web.go:167]: - ProcessWSMessages returns only an error
- process callback returns (bool, error)
- remove ProcessWSMessageResult

File: integration/util/mqtt_client.go (hunk 26 above)
Comment (e-grigorov) [reply integration/util/mqtt_client.go:74]: edge/thing/response and edge/thing/request will be placed in new private constants.

File: integration/util/web.go (hunk 31 above)
Comment (hristobojilov) [reply integration/util/web.go:131]: Any ideas here?

File: integration/util/config.go [hunk 35]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"time"
+)
+
+// TestConfiguration is a common IT configuration
+type TestConfiguration struct {
+	LocalBroker              string `env:"LOCAL_BROKER" envDefault:"tcp://localhost:1883"`
+	MqttQuiesceMs            int    `env:"MQTT_QUIESCE_MS" envDefault:"500"`
+	MqttAcknowledgeTimeoutMs int    `env:"MQTT_ACKNOWLEDGE_MS" envDefault:"3000"`
```
Comment (e-grigorov) [inline integration/util/config.go:23]: MQTT_ACKNOWLEDGE_MS -> MQTT_ACKNOWLEDGE_TIMEOUT_MS

File: integration/util/mqtt_client.go [hunk 36]
```diff
@@ -0,0 +1,114 @@
...
+		AddBroker(cfg.LocalBroker).
+		SetClientID(uuid.New().String()).
+		SetConnectTimeout(MillisToDuration(cfg.MqttConnectMs)).
+		SetKeepAlive(keepAliveTimeout).
+		SetCleanSession(true).
+		SetAutoReconnect(true)
+
+	mqttClient := MQTT.NewClient(opts)
+
+	if token := mqttClient.Connect(); token.Wait() && token.Error() != nil {
```
Comment (e-grigorov) [inline integration/util/mqtt_client.go:46]: MqttConnectMs is not used here.

Comment (e-grigorov) [reply integration/util/mqtt_client.go:46]: Just because SetConnectTimeout is used.

**Review by gboyvalenkov-bosch — APPROVED** [review]

File: integration/util/web.go [hunk 37]
```diff
@@ -0,0 +1,171 @@
...
+		var payload []byte
+		webSocketErr := websocket.Message.Receive(ws, &payload)
+		if webSocketErr != nil {
+			return fmt.Errorf("error reading from websocket: %v", webSocketErr)
+		}
+
+		envelope := &protocol.Envelope{}
+		unmarshalErr := json.Unmarshal(payload, envelope)
+		if unmarshalErr == nil {
+			finished, errProcess = process(envelope)
```
Comment (dimiter-todorov-georgiev) [inline integration/util/web.go:154]: this should be finished, err = process(envelope), and we should remove the if below. We should be able to overwrite a bad message with a good one and get a nil error if the last processed message was processed without an error.

Comment (gboyvalenkov-bosch) [reply integration/util/web.go:154]: ОК.

File: NOTICE.md [hunk 38]
```diff
@@ -58,6 +58,54 @@ apache/qpid-proton (0.36.0)
 * Project: https://github.com/apache/qpid-proton
 * Source:  https://github.com/apache/qpid-proton/tree/0.36.0
 
+github.com/caarlos0/env (1.7.0)
```
Comment (e-grigorov) [inline NOTICE.md:61]: The version is not correct, it should be: v6.10.1

File: NOTICE.md [hunk 39]
```diff
@@ -58,6 +58,54 @@ apache/qpid-proton (0.36.0)
...
+* Project: https://github.com/stretchr/testify
+* Source:  https://github.com/stretchr/testify/releases/tag/v1.8.1
+
+eclipse/paho.mqtt.golang (1.4.1)
+
+* License: Eclipse Distribution License v1.0
+* Project: https://github.com/eclipse/paho.mqtt.golang
+* Source:  https://github.com/eclipse/paho.mqtt.golang/releases/tag/v1.4.1
+
+google/uuid (1.1.1)
```
Comment (e-grigorov) [inline NOTICE.md:79]: go.mod points to version: 1.3.0

**Review by e-grigorov — APPROVED** [review]

**Pull request merged**